File: Hard/4102-minimum-threshold-path-with-limited-heavy-edges/minimum-threshold-path-with-limited-heavy-edges.cpp
class Solution{
public:
    void solve(vector<vector<int>> &e, int s, int t, vector<bool> &visited, int &high, int n){
        vector<vector<int>> adj(n);
        for(auto it: e){
            adj[it[0]].push_back(it[1]);
            adj[it[1]].push_back(it[0]);
            high=max(high, it[2]);
        }
        
        queue<int> q;
        q.push(s);
        visited[s]=true;
        while(!q.empty()){
            int t=q.front();
            q.pop();
            for(auto it: adj[t]){
                if(visited[it] == false){
                    q.push(it);
                    visited[it]=true;
                }
            }
        }
    }

    bool solve1(vector<vector<int>> &e, int s, int t, int k, int mid, int n){
        vector<vector<pair<int, int>>> adj(n);
        for(auto it: e){
            int c=(it[2]>mid?1:0);
            adj[it[0]].push_back({it[1], c});
            adj[it[1]].push_back({it[0], c});
        }

        deque<int> dq;
        vector<int> dist(n, INT_MAX);
        dist[s]=0;
        dq.push_back(s);
        while(!dq.empty()){
            int u=dq.front();
            dq.pop_front();

            for(auto it: adj[u]){
                if(it.second+dist[u]<dist[it.first]){
                    dist[it.first]=it.second+dist[u];
                    if(it.second == 0){
                        dq.push_front(it.first);
                    }
                    else{
                        dq.push_back(it.first);
                    }
                }
            }
        }

        if(dist[t]<=k)
            return true;
        return false;
    }
    
    int minimumThreshold(int n, vector<vector<int>> &e, int s, int t, int k){
        if(s == t)
            return 0;

        int low=0, high=0;
        vector<bool> visited(n, false);
        solve(e, s, t, visited, high, n);
        if(visited[t] == false)
            return -1;

        int res=high;
        while(low<=high){
            int mid=low+(high-low)/2;
            if(solve1(e, s, t, k, mid, n)){
                res=mid;
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }

        return res;
        
    }
};

File: Hard/4102-minimum-threshold-path-with-limited-heavy-edges/minimum-threshold-path-with-limited-heavy-edges.java
class Solution{
    void solve(int[][] e, int s, int t, boolean[] visited, int []high, int n){
        ArrayList<ArrayList<Integer>> adj=new ArrayList<>();
        for(int i=0; i<n; i++){
            adj.add(new ArrayList<>());
        }

        for(int[] it: e){
            adj.get(it[0]).add(it[1]);
            adj.get(it[1]).add(it[0]);
            high[0]=Math.max(high[0], it[2]);
        }
        
        Queue<Integer> q=new LinkedList<>();
        q.offer(s);
        visited[s]=true;
        while(!q.isEmpty()){
            int u=q.poll();
            for(int it: adj.get(u)){
                if(!visited[it]){
                    q.offer(it);
                    visited[it]=true;
                }
            }
        }
    }

    boolean solve1(int[][] e, int s, int t, int k, int mid, int n){
        ArrayList<ArrayList<int[]>> adj=new ArrayList<>();
        for(int i=0; i<n; i++){
            adj.add(new ArrayList<>());
        }

        for(int[] it: e){
            int c=(it[2]>mid?1:0);
            adj.get(it[0]).add(new int[]{it[1], c});
            adj.get(it[1]).add(new int[]{it[0], c});
        }

        Deque<Integer> dq=new ArrayDeque<>();
        int[] dist=new int[n];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[s]=0;
        dq.offerFirst(s);
        while(!dq.isEmpty()){
            int u=dq.pollFirst();
            for(int[] it: adj.get(u)){
                int v=it[0];
                int wt=it[1];
                if(wt+dist[u]<dist[v]){
                    dist[v]=wt+dist[u];
                    if(wt == 0){
                        dq.offerFirst(v);
                    }
                    else{
                        dq.offerLast(v);
                    }
                }
            }
        }

        if(dist[t]<=k)
            return true;
        return false;
    }
    
    public int minimumThreshold(int n, int[][] e, int s, int t, int k){
        if(s == t)
            return 0;

        int low=0;
        int[] h=new int[1];
        boolean[] visited=new boolean[n];
        solve(e, s, t, visited, h, n);
        if(!visited[t])
            return -1;

        int res=h[0];
        int high=h[0];
        while(low<=high){
            int mid=low+(high-low)/2;
            if(solve1(e, s, t, k, mid, n)){
                res=mid;
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }

        return res;
    }
}